muikku-core-plugins/src/main/resources/META-INF/resources/scripts/src/components/language-profile/body/application/cv/skill-level.tsx

```diff
@@ -0,0 +1,423 @@
+   * @param e event
+   */
+  const handleSelectChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
+    const value = e.target.value;
+    const name = e.target.name;
+
+    const updatedLanguageSkillLevels = {
+      ...languageSkillLevels,
+      [name]: value,
+    };
+
+    dispatch({
+      type: "LANGUAGE_PROFILE_UPDATE_CV_LANGUAGE",
+      payload: updatedLanguageSkillLevels,
+    } as ActionType);
+  };
+
+  /**
+   * handleRadioInputChange handles changes in the radio inputs.
+   * @param e event
+   * @param field field name
+   */
+  const handleRadioInputChange = (
+    e: React.ChangeEvent<HTMLInputElement>,
+    field: string
+  ) => {
+    const value = e.target.value;
+    const updatedLanguageSkillLevels = {
+      ...languageSkillLevels,
+      [field]: value,
+    };
+    dispatch({
+      type: "LANGUAGE_PROFILE_UPDATE_CV_LANGUAGE",
+      payload: updatedLanguageSkillLevels,
+    } as ActionType);
+  };
+
+  /**
+   * handleTextAreaChange handles changes in the text area.
+   * @param e event
+   */
+  const handleTextAreaChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
+    const value = e.target.value;
+    const name = e.target.name;
+
+    const updatedLanguageSkillLevels = {
+      ...languageSkillLevels,
+      [name]: value,
+    };
+    dispatch({
+      type: "LANGUAGE_PROFILE_UPDATE_CV_LANGUAGE",
+      payload: updatedLanguageSkillLevels,
+    } as ActionType);
+  };
+
+  /**
+   * handleUrlFieldChange handles changes in the URL input field.
+   * @param e event
+   */
+  const handleUrlFieldChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    if (validateURL(e.target.value)) {
+      setIsValidUrl(true);
+    } else {
+      setIsValidUrl(false);
+    }
+    setSampleUrl(e.target.value);
+  };
+
+  /**
+   * handleAddSampleLink adds link to sample
+   */
+  const handleAddSampleLink = () => {
+    const sample = { name: sampleName, url: sampleUrl };
+
+    const updatedLanguageSkillLevels = {
+      ...languageSkillLevels,
+      samples: [...languageSkillLevels.samples, sample],
+    };
+
+    dispatch({
+      type: "LANGUAGE_PROFILE_UPDATE_CV_LANGUAGE",
+      payload: updatedLanguageSkillLevels,
+    } as ActionType);
+
+    clearFields();
+  };
+
+  /**
+   * handleRemoveSampleLink
+   * @param index number
+   */
+  const handleRemoveSampleLink = (index: number) => {
+    const updatedLanguageSkillLevels = {
+      ...languageSkillLevels,
+    };
+    updatedLanguageSkillLevels.samples.splice(index, 1);
+
+    dispatch({
+      type: "LANGUAGE_PROFILE_UPDATE_CV_LANGUAGE",
+      payload: updatedLanguageSkillLevels,
+    } as ActionType);
+  };
+
+  React.useEffect(() => {
+    if (cv) {
+      const currentLanguageSkillLevel = cv.languages.find(
+        (l) => l.code === language.code
+      );
+      if (currentLanguageSkillLevel) {
+        setlanguageSkillLevels(currentLanguageSkillLevel);
+      }
+    }
+  }, [cv, language]);
```

Comment:
setlanguageSkillLevels logic is not used anywhere else in the component, so you can refactor this to React.memo instead which calculated value if any of dependencies have changed. This also removes unnecessary re-renders as prop change will rerender anyways